Support factories on overwritten default fields

shaffe-fr opened the pull request:
This PR allows to use the factory syntax in overwritten defaults fields.
```php
public function withGroup(): self
{
    return tap(clone $this)->overwriteDefaults([
        'group_id' => GroupFactory::new(),
    ]);
}
```

It will also work with extra attributes:
```php
$recipe = RecipeFactory::new()->create(['group_id' => GroupFactory::new()]);
```

On the current release, I had the following error when using factory:
`Error : Object of class Tests\Factories\GroupFactory could not be converted to string`.

The PR includes some tests.
Two tests are failing. It looks like it's caused by my line \r\n line endings on Windows. Do you have any idea @christophrumpel?
What do you think about this PR?

Thanks

christophrumpel commented:
Hey @shaffe-fr , thanks for this PR. I like it, but I need to check the failing test. For me it seems like a problem with the immutability feature.

shaffe-fr commented:
Hey @christophrumpel, 
Thanks for your package :)
the it_creates_factories_with_immutable_states test fails because of the creation type:
* `$recipeOne = $recipeFactory->withGroup()->make();` adds a group_id to the recipe, but it's null
* `$recipeTwo = $recipeFactory->make()` adds no group_id key.

This behaviour is caused by the line 19 of the `TranslatesFactoryData::prepareModelData` method:
`return $creationType === 'create' ? $item->create()->id : null;`.

christophrumpel commented:
Hey, so just found some time to take a look again. I have to admit I'm not 100% sure why I only returned the related item id while using the `create` method and not the `make` method. I think it was because it can get confusing when you use the `make` method and other models will be created during that call.

I tried changing this line to always return the created related item id when it is a factory and this makes all tests passing for me locally. `return $item->create()->id;`

Can you change that and push again? I will look through the current tests if we don't forget about anything.

shaffe-fr commented:
Hey, I cleaned the trait and pushed the commit. All tests are now passing :)
I removed the creationType parameter from the trait, is it okay for you?
As explained in the PR description, the it_can_add_default_states test fails on Windows because of EOL of the generated file. I think it's safe to ignore it.

shaffe-fr commented:
I have some ideas to improve the related model `with` feature:
* Add support for BelongsTo and MorphTo (failing now because they don't have a saveMany method),
* Add the possibility to create different kinds of relationship at once (only one for now because relatedModelRelationshipName is a string).

Pros of `with` compared to this PR:
- the model has his relations set,
- the foreign keys don't need to be explictly written.

Does it sound interesting to you? Can I create a new PR?

christophrumpel commented:
Hey, thanks. I only see the tests failing because of 

```
[Composer\Downloader\TransportException]                                     
  Content-Length mismatch, received 1617659 bytes out of the expected 1702304  
```
Taking a look right now.

PR to support more relations would be nice 👍I just might be picky about the naming :-)

christophrumpel commented:
Ok, I re-run the actions and everything is passing now 👍

christophrumpel commented:
Thanks for the PR 👏
